### ISSUE TITLE
RemoveAll throws if KeepAlive is false

### ISSUE THREAD

**Issue by Haukinger** [PrismLibrary/Prism#993]
### Package info

- Platform: WPF
- Prism version: 6.2.0

### Repro steps

1. Have a `MyViewModel` implement `IRegionMemberLifetime` and return `false` from `KeepAlive.get`
2. Navigate a region to a view (`_regionManager.RequestNavigate( "MyRegion" ), nameof( MyView ) )`)
3. Clear the region (`_regionManager.Regions["MyRegion"].RemoveAll()`)
4. Exception:

> "The region does not contain the specified view."
>    at Prism.Regions.Region.GetItemMetadataOrThrow(Object view)
>   at Prism.Regions.Region.Remove(Object view)
>   at Prism.Regions.Behaviors.RegionMemberLifetimeBehavior.OnActiveViewsChanged(Object sender, NotifyCollectionChangedEventArgs e)
>   at System.Collections.Specialized.NotifyCollectionChangedEventHandler.Invoke(Object sender, NotifyCollectionChangedEventArgs e)
>   at Prism.Regions.ViewsCollection.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
>   at Prism.Regions.ViewsCollection.NotifyRemove(IList items, Int32 originalIndex)
>   at Prism.Regions.ViewsCollection.RemoveFromFilteredList(Object item)
>   at Prism.Regions.ViewsCollection.SourceCollectionChanged(Object sender, NotifyCollectionChangedEventArgs e)
>   at System.Collections.Specialized.NotifyCollectionChangedEventHandler.Invoke(Object sender, NotifyCollectionChangedEventArgs e)
>   at System.Collections.ObjectModel.ObservableCollection'1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
>   at System.Collections.ObjectModel.ObservableCollection'1.RemoveItem(Int32 index)
>   at System.Collections.ObjectModel.Collection`1.Remove(T item)
>   at Prism.Regions.Region.Remove(Object view)
>   at Prism.Regions.Region.RemoveAll()

When debugging, the region's `ActiveViews` and `ItemMetadataCollection` are both empty at the time the exception is caught, while `Views` still contains an instance of `MyView`.

If `KeepAlive.get` returns `true` or `IRegionMemberLifetime` is not implemented at all, no exception is thrown when clearing the region.

**Comment by dvorn** [comment]
Seemingly relates to #382.  @Haukinger Do you have a simple project that demonstrates the problem?

**Comment by Haukinger** [comment]
Sure thing, have a look here: https://github.com/Haukinger/prism-993/tree/master

**Comment by dvorn** [comment]
This is a problem with long history:

http://compositewpf.codeplex.com/discussions/396304
http://compositewpf.codeplex.com/discussions/322324
http://compositewpf.codeplex.com/discussions/272855

Basically, the original Prism authors do not have a solution to it: in a nutshell, their recommendation is "if a  view implements KeepAlive=false, use Deactivate, otherwise you may use Remove".

Wonder whether a clean solution is possible.

**Comment by brianlagunas** [comment]
I am actually leaning towards closing this as "by-design".  Trying to get two different concepts to work the same way interchangeably is probably not a box I want to open.  If using RegionMemberLifetime, then you are explicitly using the Navigation framework.  If you are using Region.Clear/Remove, you are using View Injection.  This will be a tough sell for me.

**Comment by dvorn** [comment]
@brianlagunas However I think I can fix it...

**Comment by brianlagunas** [comment]
@dvorn that would be great!  This is a very low priority for me :)

**Comment by Haukinger** [comment]
@brianlagunas if I cannot use Region.Remove to navigate away from a view without navigating _to_ another view, is there any other way to achieve this?

**Comment by brianlagunas** [comment]
There are a number of approaches described in the links that @dvorn provided.  One of them is to deactivate the view

**Comment by dvorn** [comment]
This is the same issue as #382.  An attempt to fix it in #382 is essentially what original Prism authors recommended in http://compositewpf.codeplex.com/workitem/8224.  However, sometimes this does not work for the following reason.

A region contains views as a itemMetadataCollection, and Region.Views and Region.ActiveViews are actually ViewsCollection wrappers over itemMetadataCollection which contain a cache of filtered and sorted items from itemMetadataCollection.  ViewsCollections subscribe to changes on itemMetadataCollection so that when itemMetadataCollection is updated, the cached items are also updated.

The problem with the code 
```csharp
if (Region.Views.Contains(inactiveView)
     Region.Remove(inactiveView);
```
is that Region.Remove removes the view from itemMetadataCollection but Region.Views.Contains checks the presence of the view in the cached list.  So the result is dependent of the order in which cached collections are updated when itemMetadataCollection changes.

If ActiveViews is updated first, the RegionMemberLifetimeBehavior will be triggered and the above code executed when Views has not been updated yet which results in a crash.

The order in which Views and ActiveViews are updated depends on the order of first accesses them: they are lazy initialized, and when initialized, they subscribe to itemMetadataCollection and subscriptions are fired in the order of subscriptions.

Views/ActiveViews may be first access, that is initialized, in any order.  For instance, when a region is defined in XAML for a ContentControl, the ContentControlRegionAdapter will first access ActiveViews and then Views.  In some other cases Views may be accessed first.

Now that we understand what's happening we may think about solutions.  Many solutions are possible, including stop caching in Views or special eventing on ActiveViews to distinguish between removing a view from a region and deactivating a view.

I think the least disruptive approach would be to fix the order in which Views and ActiveViews are initialized: Views should be initialized first.  For instance, remove lazy initialization for Views and initialize them in Region's constructor.
